core/banking/fee_discount.go

```diff
@@ -0,0 +1,180 @@
+	"code.vegaprotocol.io/vega/logging"
+
+	"golang.org/x/exp/maps"
+)
+
+type partyAssetKey struct {
+	party string
+	asset string
+}
+
+func (e *Engine) feeDiscountKey(asset, party string) partyAssetKey {
+	return partyAssetKey{party: party, asset: asset}
+}
+
+func (e *Engine) RegisterTradingFees(ctx context.Context, assetID string, feesPerParty map[string]*num.Uint) {
+	updateDiscountEvents := make([]events.Event, 0, len(e.feeDiscountPerPartyAndAsset))
+
+	// ensure asset exists
+	asset, err := e.assets.Get(assetID)
+	if err != nil {
+		e.log.Panic("could not register taker fees, invalid asset", logging.Error(err))
+	}
+
+	assetQuantum := asset.Type().Details.Quantum
+
+	feesPerPartyKeys := maps.Keys(feesPerParty)
+	sort.Strings(feesPerPartyKeys)
+
+	updatedKeys := map[partyAssetKey]struct{}{}
+	for _, party := range feesPerPartyKeys {
+		fee := feesPerParty[party]
+
+		key := e.feeDiscountKey(assetID, party)
+		updatedKeys[key] = struct{}{}
+
+		if _, ok := e.feeDiscountPerPartyAndAsset[key]; !ok {
+			e.feeDiscountPerPartyAndAsset[key] = num.UintZero()
+		}
+
+		// apply decay discount amount
+		e.feeDiscountPerPartyAndAsset[key] = e.decayFeeDiscountAmount(e.feeDiscountPerPartyAndAsset[key], assetQuantum)
+
+		// add fees
+		e.feeDiscountPerPartyAndAsset[key].Add(e.feeDiscountPerPartyAndAsset[key], fee)
+
+		updateDiscountEvents = append(updateDiscountEvents, events.NewTransferFeesDiscountUpdated(
+			ctx,
+			party,
+			assetID,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
+			e.currentEpoch,
+		))
+	}
+
+	feeDiscountPerPartyAndAssetKeys := maps.Keys(e.feeDiscountPerPartyAndAsset)
+	sort.SliceStable(feeDiscountPerPartyAndAssetKeys, func(i, j int) bool {
+		if feeDiscountPerPartyAndAssetKeys[i].party == feeDiscountPerPartyAndAssetKeys[j].party {
+			return feeDiscountPerPartyAndAssetKeys[i].asset < feeDiscountPerPartyAndAssetKeys[j].asset
+		}
+
+		return feeDiscountPerPartyAndAssetKeys[i].party < feeDiscountPerPartyAndAssetKeys[j].party
+	})
+
+	for _, key := range feeDiscountPerPartyAndAssetKeys {
+		if _, ok := updatedKeys[key]; ok {
+			continue
+		}
+
+		// apply decay discount amount
+		e.feeDiscountPerPartyAndAsset[key] = e.decayFeeDiscountAmount(e.feeDiscountPerPartyAndAsset[key], assetQuantum)
+
+		updateDiscountEvents = append(updateDiscountEvents, events.NewTransferFeesDiscountUpdated(
+			ctx,
+			key.party,
+			assetID,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
+			e.currentEpoch,
+		))
+	}
+
+	e.broker.SendBatch(updateDiscountEvents)
+}
+
+func (e *Engine) ApplyFeeDiscount(ctx context.Context, asset string, party string, fee *num.Uint) (discountedFee *num.Uint, discount *num.Uint) {
+	discountedFee, discount = e.EstimateFeeDiscount(asset, party, fee)
+	if discount.IsZero() {
+		return discountedFee, discount
+	}
+
+	key := e.feeDiscountKey(asset, party)
+	defer e.broker.Send(
+		events.NewTransferFeesDiscountUpdated(ctx,
+			party, asset,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
```

Comment:
What's the intention here ? Send an event with the fee value before the subtraction, or after the subtraction ?

If that's the latter, then I believe the evaluation of `e.feeDiscountPerPartyAndAsset[key].Clone()` is too early. You will have to wrap this in a function.

Reply:
fixed